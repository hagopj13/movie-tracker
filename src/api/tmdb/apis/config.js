// @flow
import axios from '../axiosInstance';

export const getConfig = () => axios.get('/configuration');

export const getAllGenres = () => axios.get('/genre/movie/list');
