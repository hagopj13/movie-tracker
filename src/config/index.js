import React from 'react';
import DiscoverIcon from '@material-ui/icons/Movie';
import UpcomingIcon from '@material-ui/icons/Update';
import ProfileIcon from '@material-ui/icons/Person';

export const navigationItems = [
  {
    id: 'DISCOVER',
    label: 'Discover',
    to: '/',
    icon: <DiscoverIcon />,
    requiresAuth: false,
  },
  {
    id: 'UPCOMING',
    label: 'Upcoming',
    to: '/upcoming',
    icon: <UpcomingIcon />,
    requiresAuth: false,
  },
  {
    id: 'PROFILE',
    label: 'Profile',
    to: '/profile',
    icon: <ProfileIcon />,
    requiresAuth: true,
  },
];

export const dialogTypes = {
  LOGIN: 'LOGIN',
};

export const sortOptions = [
  { value: 'popularity.desc', label: 'Popularity Descending' },
  { value: 'popularity.asc', label: 'Popularity Ascending' },
  { value: 'vote_average.desc', label: 'Rating Descending' },
  { value: 'vote_average.asc', label: 'Rating Ascending' },
  { value: 'release_date.desc', label: 'Release Date Descending' },
  { value: 'release_date.asc', label: 'Release Date Ascending' },
];
