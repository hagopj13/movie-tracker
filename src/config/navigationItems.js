import React from 'react';
import DiscoverIcon from '@material-ui/icons/Movie';
import UpcomingIcon from '@material-ui/icons/Update';
import ProfileIcon from '@material-ui/icons/Person';

export default [
  {
    id: 'DISCOVER',
    label: 'Discover',
    to: '/',
    icon: <DiscoverIcon />,
    requiresAuth: false,
  },
  {
    id: 'UPCOMING',
    label: 'Upcoming',
    to: '/upcoming',
    icon: <UpcomingIcon />,
    requiresAuth: false,
  },
  {
    id: 'PROFILE',
    label: 'Profile',
    to: '/profile',
    icon: <ProfileIcon />,
    requiresAuth: true,
  },
];
