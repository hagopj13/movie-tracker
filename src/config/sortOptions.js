export default [
  { value: 'popularity.desc', label: 'Popularity Descending' },
  { value: 'popularity.asc', label: 'Popularity Ascending' },
  { value: 'vote_average.desc', label: 'Rating Descending' },
  { value: 'vote_average.asc', label: 'Rating Ascending' },
  { value: 'release_date.desc', label: 'Release Date Descending' },
  { value: 'release_date.asc', label: 'Release Date Ascending' },
];
