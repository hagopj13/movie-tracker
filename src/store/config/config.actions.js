// @flow
import type { ImagesConfig, Genre } from 'types';

import ConfigActionTypes from './config.types';

const fetchImagesConfig = () => ({
  type: ConfigActionTypes.FETCH_IMAGES_CONFIG,
});

const fetchImagesConfigStart = () => ({
  type: ConfigActionTypes.FETCH_IMAGES_CONFIG_START,
});

const fetchImagesConfigSuccess = (imagesConfig: ImagesConfig) => ({
  type: ConfigActionTypes.FETCH_IMAGES_CONFIG_SUCCESS,
  payload: { imagesConfig },
});

const fetchImagesConfigFailure = (error: string) => ({
  type: ConfigActionTypes.FETCH_IMAGES_CONFIG_FAILURE,
  payload: { error },
});

const fetchAllGenres = () => ({
  type: ConfigActionTypes.FETCH_ALL_GENRES,
});

const fetchAllGenresStart = () => ({
  type: ConfigActionTypes.FETCH_ALL_GENRES_START,
});

const fetchAllGenresSuccess = (genres: Genre[]) => ({
  type: ConfigActionTypes.FETCH_ALL_GENRES_SUCCESS,
  payload: { genres },
});

const fetchAllGenresFailure = (error: string) => ({
  type: ConfigActionTypes.FETCH_ALL_GENRES_FAILURE,
  payload: { error },
});

export default {
  fetchImagesConfig,
  fetchImagesConfigStart,
  fetchImagesConfigSuccess,
  fetchImagesConfigFailure,
  fetchAllGenres,
  fetchAllGenresStart,
  fetchAllGenresSuccess,
  fetchAllGenresFailure,
};
