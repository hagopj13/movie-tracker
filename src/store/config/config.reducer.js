// @flow
import { handleActions } from 'redux-actions';

import type { ImagesConfig, Genre } from 'types';

import ConfigActionTypes from './config.types';

type State = {
  images: ImagesConfig,
  genres: Genre[],
};

export const initialState: State = {
  images: {
    baseImageUrl: '',
    backdropSize: '',
    posterSize: '',
    profileSize: '',
  },
  genres: [],
};

const fetchImagesConfigSuccess = (state: State, action: any) => ({
  ...state,
  images: action.payload.imagesConfig,
});

const fetchAllGenresSuccess = (state: State, action: any) => ({
  ...state,
  genres: action.payload.genres,
});

const actionHandler = {
  [ConfigActionTypes.FETCH_IMAGES_CONFIG_SUCCESS]: fetchImagesConfigSuccess,
  [ConfigActionTypes.FETCH_ALL_GENRES_SUCCESS]: fetchAllGenresSuccess,
};

export default handleActions(actionHandler, initialState);
