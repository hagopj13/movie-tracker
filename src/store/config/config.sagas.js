import { takeLatest, all, call, put } from 'redux-saga/effects';

import * as api from 'api/tmdb';
import { convertResponseToImagesConfig } from 'api/tmdb/utils';

import ConfigActionTypes from './config.types';
import configActions from './config.actions';

export function* fetchImagesConfig() {
  yield put(configActions.fetchImagesConfigStart());
  try {
    const { data } = yield call(api.getConfig);
    const imagesConfig = yield call(convertResponseToImagesConfig, data);
    yield put(configActions.fetchImagesConfigSuccess(imagesConfig));
  } catch (error) {
    yield put(configActions.fetchImagesConfigFailure(error.status_message));
  }
}

export function* fetchAllGenres() {
  yield put(configActions.fetchAllGenresStart());
  try {
    const { data } = yield call(api.getAllGenres);
    yield put(configActions.fetchAllGenresSuccess(data.genres));
  } catch (error) {
    yield put(configActions.fetchAllGenresFailure(error.status_message));
  }
}

export function* onFetchImagesConfig() {
  yield takeLatest(ConfigActionTypes.FETCH_IMAGES_CONFIG, fetchImagesConfig);
}

export function* onFetchAllGenres() {
  yield takeLatest(ConfigActionTypes.FETCH_ALL_GENRES, fetchAllGenres);
}

export default function* configSagas() {
  yield all([call(onFetchImagesConfig), call(onFetchAllGenres)]);
}
