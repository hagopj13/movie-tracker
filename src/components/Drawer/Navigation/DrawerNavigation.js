// @flow
import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { useLocation, NavLink } from 'react-router-dom';
import MenuList from '@material-ui/core/MenuList';
import MenuItem from '@material-ui/core/MenuItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';

import authSelectors from 'store/auth/auth.selectors';
import navigationItems from 'config/navigationItems';

type Props = {
  isAuth: boolean,
};

export const DrawerNavigation = (props: Props) => {
  const { isAuth } = props;

  const location = useLocation();

  const isActiveRoute = (path: string) => location.pathname === path;

  return (
    <MenuList>
      {navigationItems.map(
        (navItem) =>
          (isAuth || !navItem.requiresAuth) && (
            <NavLink
              key={navItem.id}
              to={navItem.to}
              style={{ textDecoration: 'none', color: 'inherit' }}
            >
              <MenuItem selected={isActiveRoute(navItem.to)}>
                <ListItemIcon>{navItem.icon}</ListItemIcon>
                <ListItemText primary={navItem.label} />
              </MenuItem>
            </NavLink>
          ),
      )}
    </MenuList>
  );
};

const mapStateToProps = createStructuredSelector({
  isAuth: authSelectors.selectIsAuth,
});

export default connect(mapStateToProps)(DrawerNavigation);
