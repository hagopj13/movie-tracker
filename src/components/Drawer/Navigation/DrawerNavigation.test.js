import React from 'react';
import { useLocation, NavLink } from 'react-router-dom';
import { createShallow } from '@material-ui/core/test-utils';
import MenuItem from '@material-ui/core/MenuItem';

import { navigationItems } from 'config';

import { DrawerNavigation } from './DrawerNavigation';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: jest.fn(),
}));

describe('DrawerNavigation component', () => {
  let shallow;
  let wrapper;

  beforeAll(() => {
    shallow = createShallow();
  });

  beforeEach(() => {
    useLocation.mockReturnValue({ pathname: '/' });
    const mockProps = {
      isAuth: true,
    };
    wrapper = shallow(<DrawerNavigation {...mockProps} />);
  });

  it('should render the DrawerNavigation component correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render NavLink for each navigation item', () => {
    expect(wrapper.find(NavLink)).toHaveLength(navigationItems.length);
  });

  it('should set selected prop of MenuItem to true when it matched the current location', () => {
    expect(wrapper.find(MenuItem).at(0).prop('selected')).toBe(true);
  });

  it('should only render navigation items that do no require auth if isAuth is false', () => {
    const mockProps = {
      isAuth: false,
    };
    const newWrapper = shallow(<DrawerNavigation {...mockProps} />);
    const publicNavItems = navigationItems.filter((navItem) => !navItem.requiresAuth);
    expect(newWrapper.find(NavLink)).toHaveLength(publicNavItems.length);
  });
});
