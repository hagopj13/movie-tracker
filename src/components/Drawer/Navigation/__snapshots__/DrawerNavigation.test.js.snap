// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawerNavigation component should render the DrawerNavigation component correctly 1`] = `
<ForwardRef(MenuList)>
  <NavLink
    key="DISCOVER"
    style={
      Object {
        "color": "inherit",
        "textDecoration": "none",
      }
    }
    to="/"
  >
    <WithStyles(ForwardRef(MenuItem))
      selected={true}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <MovieIcon />
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Discover"
      />
    </WithStyles(ForwardRef(MenuItem))>
  </NavLink>
  <NavLink
    key="UPCOMING"
    style={
      Object {
        "color": "inherit",
        "textDecoration": "none",
      }
    }
    to="/upcoming"
  >
    <WithStyles(ForwardRef(MenuItem))
      selected={false}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <UpdateIcon />
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Upcoming"
      />
    </WithStyles(ForwardRef(MenuItem))>
  </NavLink>
  <NavLink
    key="PROFILE"
    style={
      Object {
        "color": "inherit",
        "textDecoration": "none",
      }
    }
    to="/profile"
  >
    <WithStyles(ForwardRef(MenuItem))
      selected={false}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <PersonIcon />
      </WithStyles(ForwardRef(ListItemIcon))>
      <WithStyles(ForwardRef(ListItemText))
        primary="Profile"
      />
    </WithStyles(ForwardRef(MenuItem))>
  </NavLink>
</ForwardRef(MenuList)>
`;
